[improve][client] Batch receives the same topic messages to support c…

congbobo184 opened the pull request:
…umulative ack

<!--
### Contribution Checklist
  
  - PR title format should be *[type][component] summary*. For details, see *[Guideline - Pulsar PR Naming Convention](https://docs.google.com/document/d/1d8Pw6ZbWk-_pCKdOmdvx9rnhPiyuxwq60_TrD68d7BA/edit#heading=h.trs9rsex3xom)*. 

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.
-->

<!-- Either this PR fixes an issue, -->

Fixes #xyz

<!-- or this PR is one task of an issue -->

Master Issue: #xyz

### Motivation

<!-- Explain here the context, and why you're making that change. What is the problem you're trying to solve. -->

### Modifications

<!-- Describe the modifications you've done. -->

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] Anything that affects deployment

### Documentation

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

- [ ] `doc` <!-- Your PR contains doc changes. Please attach the local preview screenshots (run `sh start.sh` at `pulsar/site2/website`) to your PR description, or else your PR might not get merged. -->
- [ ] `doc-required` <!-- Your PR changes impact docs and you will update later -->
- [ ] `doc-not-needed` <!-- Your PR changes do not impact docs -->
- [ ] `doc-complete` <!-- Docs have been already added -->

### Matching PR in forked repository

PR in forked repository: <!-- ENTER URL HERE -->

<!--
After opening this PR, the build in apache/pulsar will fail and instructions will
be provided for opening a PR in the PR author's forked repository.

apache/pulsar pull requests should be first tested in your own fork since the 
apache/pulsar CI based on GitHub Actions has constrained resources and quota.
GitHub Actions provides separate quota for pull requests that are executed in 
a forked repository.

The tests will be run in the forked repository until all PR review comments have
been handled, the tests pass and the PR is approved by a reviewer.
-->

Commit:
[improve][client] Batch receives the same topic messages to support cumulative ack

eolivelli reviewed (changes requested):
This is going to reduce performances of applications that consume from a partitioned topic, the application will receive smaller batches.

Commit:
Merge branch 'master' into client/cumulative_ack_for_parition_topic

Commit:
[improve][client] Batch receives the same topic messages to support cumulative ack

Commit:
Merge remote-tracking branch 'apache/master' into client/cumulative_ack_for_parition_topic

Commit:
fix some comments

Commit:
fix some comments

Commit:
fix some comments

0xflotus committed:
[fix][doc] typo in functions-runtime (#18976)

aloyszhang committed:
[fix][broker]fix multi invocation for ledger createComplete (#18975)

### Motivation

fix multi invocation for ledger `createComplete`


### Modifications

Only call `createComplete` at the point of creating ledger timeout if the ledger is not created normally

coderzc committed:
[feat][broker][PIP-195] Implement BookkeeperBucketSnapshotStorage - part3 (#17677)

AnonHxy committed:
[fix][client] Fix too many redelivered messages when one message acked timeout in listener (#18911)

codelipenghui committed:
[improve][doc] Update PR template and PIP for metrics changes (#18961)

asafm committed:
[fix][doc] Ammend PIP proposal around discussions  (#18974)

poorbarcode committed:
[fix] [broker] getLastMessageId returns a wrong batch index of last message if enabled read compacted (#18877)

### Motivation

The method `consumer.getLastMessageId` will return the latest message which can be received.
- If disabled `read compacted`, will return the last confirmed position of `ManagedLedger`.
- If enabled `read compacted`, will return the latest message id which can be read from the compacted topic.

If we send a batch message like this:

```java
producer.newMessage().key("k1").value("v0").sendAsync(); // message-id is [3:1,-1:0]
producer.newMessage().key("k1").value("v1").sendAsync(); // message-id is [3:1,-1:1]
producer.newMessage().key("k1").value("v2").sendAsync(); // message-id is [3:1,-1:2]
producer.newMessage().key("k2").value("v0").sendAsync(); // message-id is [3:1,-1:3]
producer.newMessage().key("k2").value("v1").sendAsync(); // message-id is [3:1,-1:4]
producer.newMessage().key("k2").value(null).sendAsync(); // message-id is [3:1,-1:5]
producer.flush();
```

After the compaction task is done, the messages with key `k2` will be deleted by the compaction task. Then the latest message that can be received will be `[3:1:-1:2]`.

--- 
When we call `consumer.getLastMessageId`, the expected result is:

```
[3:1,-1:2]
```

--- 
But the actual result is:

```
[3:1,-1:5]
```

### Modifications
If enabled `read compacted` and the latest entry of the compacted topic is a batched message, extract the entry and calculate all internal messages, then return the latest message which is not marked `compacted out`.

pen4 committed:
[fix][sec] upgrade org.testng:testng to 7.7.0 (#18977)

michaeljmarshall committed:
[fix][broker] Copy proto command fields into final variables in ServerCnx (#18987)

### Motivation

In the `PulsarDecoder`, we use a single `BaseCommand` object and overwrite it for each incoming protocol message. As a result, it is not safe to publish any references to a proto command to other threads.

Here is the single `BaseCommand`:

https://github.com/apache/pulsar/blob/82237d3684fe506bcb6426b3b23f413422e6e4fb/pulsar-common/src/main/java/org/apache/pulsar/common/protocol/PulsarDecoder.java#L99

Here is the method call that resets the object:

https://github.com/apache/pulsar/blob/82237d3684fe506bcb6426b3b23f413422e6e4fb/pulsar-common/src/main/java/org/apache/pulsar/common/protocol/PulsarDecoder.java#L114

Note that the call to `parseFrom` first calls `clear()`, which resets all values on the object.

This PR copies relevant values or objects into other variables.

### Modifications

* Replace `command` with `tcId` since the latter is a final variable meant to be published to another thread.
* Move logic to copy certain command fields to earlier in method for `handleSubscribe`
* Copy `ack` object to new `CommandAck` when there is a broker interceptor. Note that copying this command is likely somewhat costly, so we only do it when there is an interceptor configured.

### Verifying this change

This is a trivial change that is already covered by tests.

### Documentation

- [x] `doc-not-needed` 

This is an internal change.

### Matching PR in forked repository

PR in forked repository: https://github.com/michaeljmarshall/pulsar/pull/8

liangyepianzhou committed:
[fix][doc]Fix indentation problem in docker-compose.yaml (#18994)

nicoloboschi committed:
[fix][broker] Fix deadlock in PendingAckHandleImpl (#18989)

nicoloboschi committed:
[improve][io] reduce dependencies of pulsar-io-common module (#19000)

lhotari committed:
[improve][broker] Add javadoc for PulsarDecoder handle* method parameter lifecycle (#18998)

lhotari committed:
[fix][test] Fix flaky test PulsarMultiListenersWithoutInternalListenerNameTest.setup (#19006)

lhotari committed:
[improve][broker] Omit making a copy of CommandAck when there are no broker interceptors (#18997)

lhotari committed:
[fix][broker] Fix direct memory leak in RawReaderImpl (#18928)

lhotari committed:
[fix][test] Run BlobStoreManagedLedgerOffloader* tests separately and reduce memory consumption (#19009)

lhotari committed:
[fix][broker] Fix thread-safety issue with currentLedgerEntries and currentLedgerSize fields (#17868)

codelipenghui committed:
[improve][broker] Add logs for why namespace bundle been split (#19003)

heesung-sohn committed:
[improve][broker][client] PIP-215 Added TopicCompactionStrategy for StrategicTwoPhaseCompactor and TableView. (#18195)

tisonkun committed:
[feat][io] Upgarde ClickHouse driver to support loadbalance policy (#18774)

Signed-off-by: tison <wander4096@gmail.com>

gaozhangmin committed:
[improve][admin]Remove repeating validateTopicOwnershipAsync action (#18996)

Co-authored-by: gavingaozhangmin <gavingaozhangmin@didiglobal.com>

lhotari committed:
[improve][io] Upgrade Kafka client, connect runtime to 2.8.2 and Confluent version to 6.2.8 (#19010)

lhotari committed:
[improve][broker] Use validateTopicOperationAsync methods in all cases (#19016)

coderzc committed:
[feat][broker][PIP-195] Implement delayed message bucket snapshot recover - part5 (#18420)

aahmed-se committed:
[improve] Add PulsarExceptionBase class that supports slf4j like parameterized string formatting (#18968)

Co-authored-by: Ali Ahmed <alia@splunk.com>

michaeljmarshall committed:
[improve][broker] Only get consumer future when interceptor not null (#19022)

tuteng committed:
[fix][broker]Update rest api response class (#18965)

thetumbled committed:
[fix][txn] Correct the prompt message (#17009)

Co-authored-by: fengwenzhi <fengwenzhi.max@bigo.sg>

Anonymitaet committed:
[feat][doc] add Pulsar API content (#19002)

codelipenghui committed:
Update .asf.yaml to avoid merge PRs without approvals to the active release branches (#19024)

lhotari committed:
[fix][broker] Fix thread safety issue in info-internal admin api for partitioned topics (#19021)

codelipenghui committed:
Revert "Update .asf.yaml to avoid merge PRs without approvals to the active release branches (#19024) (#19034)

Commit:
[improve][io] Debezium sources: Support loading config from secrets (#19004)

Commit:
Merge remote-tracking branch 'apache/master' into client/cumulative_ack_for_parition_topic
